src/function.cpp

```diff
@@ -0,0 +1,108 @@
+#include "function.h"
+#include "utils.h"
+
+#include <algorithm>
+#include <map>
+#include <vector>
+
+using namespace std;
+using namespace smt;
+
+namespace {
+map<string, DeclaredFunction, std::less<>> calleeMap;
+}
+
+DeclaredFunction::DeclaredFunction(vector<mlir::Type> &&domain,
+                                   mlir::Type &&range, FnDecl &&decl)
+    : domain(move(domain)), range(move(range)), decl(move(decl)) {}
+
+DeclaredFunction DeclaredFunction::declare(std::vector<mlir::Type> &&domain,
+                                           mlir::Type &&range,
+                                           const std::string_view name) {
+  // analyze verification complexity
+  auto verifComplexity = Complexity::SCALAR;
+  for (const auto &operand_ty : domain) {
+    if (operand_ty.isa<mlir::TensorType>()) {
+      verifComplexity = Complexity::TENSOR;
+    } else if (operand_ty.isa<mlir::MemRefType>()) {
+      verifComplexity = Complexity::MEMREF;
+    } else if (!operand_ty.isIntOrIndexOrFloat()) {
+      throw UnsupportedException("Invalid operand type");
+    }
+  }
```

Comment:
1. It is not clear when a function both takes tensor and memref. Or is it assumed that such case cannot happen?

2. What about simply defining two bool variables: `bool hasMemRef` and `bool hasTensor`, and marking a flag to true if any of domain/range is Tensor/MemRef type? I think this will simplify this PR.

Reply:
1. My first insight was that verifying memref(s) will always be more 'complex' than verifying tensor(s), so a function that takes both memref and tensor would be considered as 'memref function'. But now that I think back on it again, it seems that verifying memref and verifying tensor two different tasks, and one's difficulty doesn't necessarily cover the other.
2. Agreed, for the aforementioned reasons. But isn't it only the domain types that affect the difficulty? According to our assumption on UF, we'll get the same value for the same input and different value otherwise, regardless of the range type.